src/Markdown/CommonMark/ExternalLinkRenderer.php

```diff
@@ -136,7 +141,47 @@ public function render(Node $node, ChildNodeRendererInterface $childRenderer): H
 
                     return new HtmlElement('div');
                 }
+            } elseif (!$isApRequest && null === $apActivity) {
+                if ($user = $this->userRepository->findOneBy(['apProfileId' => $url])) {
+                    $this->logger->debug('found user for URL {u}', ['u' => $url]);
+                    // @admin@mbin.tld
+                    $name = $user->username;
+                    // admin
+                    $name2 = $this->mentionManager->getUsername($user->username);
+                    // @admin
+                    $name3 = '@'.$this->mentionManager->getUsername($user->username);
+                    // we need that whole checking here because we do not just get the raw text of the link, but instead only the first half
+                    // so if the raw text is [@admin@mbin.tld](https://mbin.tld/u/admin) we get "admin"
+                    if ($childContent === $name || $childContent === $name2 || $childContent === $name3) {
+                        return new HtmlElement('span', contents: $this->renderUser($user));
+                    } else {
+                        $this->logger->debug('found user for URL {u}, but the link text "{t}" is not the same as the username "{name}" or "{name2}" or "{name3}"', ['u' => $url, 't' => $childContent, 'name' => $name, 'name2' => $name2, 'name3' => $name3]);
+                    }
+                } elseif ($magazine = $this->magazineRepository->findOneBy(['apProfileId' => $url])) {
+                    $this->logger->debug('found magazine for URL {u}', ['u' => $url]);
+                    // technology@lemmy.world
+                    $name = $magazine->name;
+                    if (!str_starts_with($name, '@') && !str_starts_with($name, '!')) {
+                        // @technology@lemmy.world
+                        $name = "@$name";
+                    }
+                    // technology
+                    $name2 = $this->mentionManager->getUsername($name);
+                    // !technology
+                    $name3 = '!'.$this->mentionManager->getUsername($name);
```

Comment:
You're calling `his->mentionManager->getUsername` twice, which sounds inefficient.

Reply:
Same for above on line 150.